### DISCUSSION TITLE
Converting Ultisnips to Luasnips

### DISCUSSION THREAD

**Discussion by L3MON4D3** [L3MON4D3/LuaSnip#545] [General]
This discussion is for collecing general advice and answering questions on converting Ultisnip-snippets to luasnip.
If you have a question, advice, or anything else related really, post it to this thread :D

**Comment by evesdropper** [comment]
Alright then, starting my megathread post here :eyes:

## Progress/Advice/Resources
I've converted a majority of my snippets from UltiSnips to LuaSnip and hope to finish the rest ASAP as school is starting and I simply would shudder at the thought of having to create a matrix from scratch for my linear algebra class. Will update as I take notes and see what I still need to turn into snippets.

### Takeaways/Advice
- A majority of snippets can be converted with some basic knowledge of how LuaSnips works (insert nodes, text nodes, knowledge of triggers, etc).
    - Start converting! Look through some resources, create 1-2 small snippets that work, and then convert said snippets into Lua snippets to save some time. (e.g. something like [this](https://github.com/evesdropper/dotfiles/blob/main/.config/nvim/luasnip/lua.lua) but probably more well documented)
    - My workflow is like so: learning something (start with basic text/insert and autoexpand, then math mode, then function nodes, then eventually I want to get to dynamic nodes), converting all the snippets, and then repeat.
    - Use `fmt` - it makes your snippets a lot neater and easier to read/convert. The only exception is when you only use very few nodes (e.g 1 text node).

- How do I convert [something] from UltiSnips to LuaSnip?
This is a bit of a work in progress as I am also trying to figure it out myself, but here is what I have so far:
    - `A` snippets: Auto-expand snippets are included in a separate list. So something like `return { regular snippets, autosnippets}` and setting `ls.config.set_config({enable_autosnippets = true})` somewhere in your config will be fine.
    - Snippets in math mode: Conditional snippets (Set the third argument to `condition=math`) with the following:
```Lua
local function math():
    return vim.api.nvim_eval('vimtex#syntax#in_mathzone()') == 1
end
```
Unfortunately, I haven't figured out how to do this for the rest of the modes (see questions), but we'll get there hopefully.
    - Pre-expand snippets (e.g. tables, matrices): Use dynamic nodes. It might take a while to implement them; @L3MON4D3 has some great [examples](https://github.com/L3MON4D3/Dotfiles/blob/master/.config/nvim/luasnippets/tex.lua) that use user input; I opted to create [similar snippets](https://github.com/evesdropper/dotfiles/blob/8a9bb2dedaeb3ccbc88612c6e927ca3aa8c92e8f/.config/nvim/luasnip/tex.lua#L353) that rely on regex triggers as those are more native to me.

### Resources 

- TJ DeVries's LuaSnips Videos ([Intro](https://www.youtube.com/watch?v=Dn800rlPIho) and [advanced](https://www.youtube.com/watch?v=KtQZRAkgLqo)).
- Ziontee113's [Ultimate Guide Video](https://www.youtube.com/watch?v=ub0REXjhpmk)
I personally found the Ultimate Guide to be a better soft introduction (i.e. something I could watch while biking and kind of refresh my knowledge on LuaSnips), but TJ's advanced video seemed to be better in my opinion for the more complex snippet examples.
- [Wiki](https://github.com/L3MON4D3/LuaSnip/wiki) and [Docs](https://github.com/L3MON4D3/LuaSnip/blob/master/DOC.md) are also great resources, but in my opinion, they make a bit more sense after you have a bit more knowledge about snippets. The same goes for any LuaSnips dotfiles you may find.

## Questions
Still a few things I hope to be able to understand more/be able to convert. Feel free to tag me and let me know if you have implemented the following!

- `b` snippets - snippets that start at the beginning of a line.
- ~~`w` snippets - snippets that trigger with word boundaries.~~ I cannot read; `wordTrig` is true by default.
I know both should be conditions, but defining said conditions is slightly difficult for me.
- Integrating VimTeX with environments (e.g. my goal is to build up a list of TikZ snippets over time, and I want to make sure my snippets only activate in TikZ.
- Advanced Regex triggers: I want to make sure I can trigger stuff like `log` -> `\log` and such without making too many snippets for myself. I started creating a table and triggering a function node to change `trig` -> `\full word` only to realize that Lua patterns do not support alternation :laughing:
- Visual Mode: some UltiSnips snippets accept visual highlights as arguments, and I want to be able to do that too in LuaSnip. I looked on Reddit and it definitely is possible, but there was not much more further information or examples.

Converting snippets from UltiSnips to LuaSnip is quite a task; however, LuaSnip is packed with features and I'm sure the journey will be worth it. If you need some references, you can find some of my snippets [here](https://github.com/evesdropper/dotfiles/blob/main/.config/nvim/luasnip/tex.lua) - hopefully they will be helpful!

**Reply by L3MON4D3** [reply]
Woah awesome, now that's detailed :+1::+1:
Your experiences with luasnip is interesting, it's very useful to see which parts you couldn't find info about/were harder to understand.

Regarding the questions:
* `b`: the condition for the beginning of the line is actually included in luasnip! You can do
  ```lua
  condition = require("luasnip.extras.expand_conditions").line_begin
  ```
  (the source for that is [here](https://github.com/L3MON4D3/LuaSnip/blob/b7d520e1c6edf8e3eb50f162e79bcc6c8c3f38a5/lua/luasnip/extras/expand_conditions.lua#L3-L6))

* environments: A really cool way here would be to utilize treesitter to figure out what environments one is currently in, and expand based on that!

  There's #339, but it doesn't seem like they got around to implementing something as of yet.
  OTOH I'm a bit more comfortable with treesitter now, might be I find some time to look into this.
  Maybe there are also already plugins for just this.

* Advanced regex: Yeah, that's a bit of an unnecessary limitation, I think we'll introduce real regexes (either ECMAScript via jsregexp or vim-regex).

* Visual mode: Ah, for this `store_selection_keys` must be set in `ls.config.setup` ([example](https://github.com/L3MON4D3/Dotfiles/blob/b974253b38a53768f5a89e965b4f3b6744f65e5c/.config/nvim/lua/plugins/luasnip/init.lua#L12)), pressing those in `VISUAL` will populate the env-variables `SELECT_DEDENT`, `TM_SELECTED_TEXT` and `SELECT_RAW` (better avoid `TM_SELECTED_TEXT`, it's designed to be compatible with vscode, and has very quirky(stupid) behaviour because of that)

**Reply by evesdropper** [reply]
Thanks for the responses! I'll definitely find the visual mode and `b` snippets useful to implement a few more snippets. Regex is not really a big issue since Lua patterns can get the job done for the most part. Perhaps there could be a feature which allows either full regex for larger patterns and Lua patterns for simpler examples (e.g. `regTrig = full; regTrig = lua` or something, I'm not sure)

As for environments, I use VimTeX for syntax highlighting, so Treesitter is disabled. :frowning: On the positive side, I believe VimTeX makes it a bit easier to look through syntax. Previously, something that works for environments in UltiSnips is the following, but I'm having trouble converting it to Lua.

```python
# use context(env(<name>))
def env(name):
    [x,y] = vim.eval("vimtex#env#is_inside('" + name + "')") 
    return x != '0' and y != '0'
```

Once this can be converted, I think this could be expanded just a bit - if the environment can be determined, I wonder if it's possible to determine the subenvironment (e.g. for `minted`, I want to add some Python code; let LuaSnip detect that it's `minted/python`, and use Python snippets there). Of course, that's not a big issue as adding some code won't be a huge part of my LaTeX typesetting, but perhaps an idea that I could explore in the future. I'd definitely need to look through VimTeX syntax for more information.

**Reply by L3MON4D3** [reply]
> Perhaps there could be a feature which allows either full regex for larger patterns and Lua patterns for simpler examples (e.g. regTrig = full; regTrig = lua or something, I'm not sure)

Mhmm, so far I've though of adding two keys, one for ecmascript-regex, another for vim-regex, but replacing (backwards-compatibly :D) `regTrig` with something like `trigger_type` $\in$ `{"plain", "lua", "ecma", "vim"}`, default `"plain"` would be much more readable (and make up for the choice of `regTrig` for a pattern :sweat_smile:).


> As for environments, I use VimTeX for syntax highlighting, so Treesitter is disabled. :frowning: On the positive side, I believe VimTeX makes it a bit easier to look through syntax. Previously, something that works for environments in UltiSnips is the following, but I'm having trouble converting it to Lua.
> 
> ```python
> # use context(env(<name>))
> def env(name):
>     [x,y] = vim.eval("vimtex#env#is_inside('" + name + "')") 
>     return x != '0' and y != '0'
> ```

Hmm, that looks doable, I guess you know about `vim.cmd` and `nvim_exec`? Maybe they have problems returning multiple values?

> Once this can be converted, I think this could be expanded just a bit - if the environment can be determined, I wonder if it's possible to determine the subenvironment (e.g. for `minted`, I want to add some Python code; let LuaSnip detect that it's `minted/python`, and use Python snippets there).

Just for reference, `ft_func` can be customized for this (see readme), you can basically set the function that luasnip uses for determining the currently valid filetypes.

But honestly, that sounds very much like a job for treesitter, just out of interest, what's keeping you from it? (I also write a bit of latex, maybe I'm missing out :D)

**Reply by leiserfg** [reply]
> Mhmm, so far I've though of adding two keys, one for ecmascript-regex, another for vim-regex, but replacing (backwards-compatibly :D) regTrig with something like trigger_type {"plain", "lua", "ecma", "vim"}, default "plain" would be much more readable (and make up for the choice of regTrig for a pattern sweat_smile).

Sadly nvim does not expose a way to get the matches of the regex, only the full match. So if we wanna make it, a PR to nvim will be required. I assume if someone wants complex regexes, probably (s)he will want to manipulate the string as well.

**Reply by evesdropper** [reply]
> Mhmm, so far I've though of adding two keys, one for ecmascript-regex, another for vim-regex, but replacing (backwards-compatibly :D) regTrig with something like trigger_type  {"plain", "lua", "ecma", "vim"}, default "plain" would be much more readable (and make up for the choice of regTrig for a pattern :sweat_smile:)

Looking forward to that - sounds like a great idea! It would definitely save a bit of time. On that note, would this trigger type be possible in postfix? I realize that perhaps postfix would be the more ideal way to work with greek letters/some other math symbols.

> But honestly, that sounds very much like a job for treesitter, just out of interest, what's keeping you from it? (I also write a bit of latex, maybe I'm missing out :D)

Hmm, I guess I'm a bit used to built-in syntax highlighting with VimTeX + the VimTeX syntax patterns, so it is quite difficult for me to move over. I think I saw some Treesitter detect math mode a while back that took up a bit more implementation so I'd say VimTeX syntax is probably better as of right now.

As for the function, I think I have something that might work (at least theoretically), but it errors.

```Lua
local function env(name)
    x, y =  vim.api.nvim_eval("vimtex#env#is_inside('" .. name .. "')")
    return x ~= 0 and y ~= 0

-- some local tikz implementation
```


> Hmm, that looks doable, I guess you know about `vim.cmd` and `nvim_exec`? Maybe they have problems returning multiple values?

I'm not sure what the problem is; currently, I get some errors about indexing, so I would assume so. I'm not quite sure how to resolve this right now though. Perhaps in the future.

**Reply by L3MON4D3** [reply]
Opened an issue for more trigger-types, #547

>would this trigger type be possible in postfix?

I think so, Iirc postfix also work with `regTrig` (those two concepts are orthogonal: postfix just finds the word before the trigger, this happens entirely after the trigger is matched)

> Hmm, I guess I'm a bit used to built-in syntax highlighting with VimTeX + the VimTeX syntax patterns, so it is quite difficult for me to move over. I think I saw some Treesitter detect math mode a while back that took up a bit more implementation so I'd say VimTeX syntax is probably better as of right now.

Ahh yeah I get that :D
I'll have to take a look at vimtex some time, treesitter can be a bit funky (color changes when the cursoroves over it xD)

> I realize that perhaps postfix would be the more ideal way to work with greek letters/some other math symbols.

BTW, interesting, how so?

**Reply by evesdropper** [reply]
Woops, school got the best of me this week, but at least taking notes was pretty easy with LuaSnips (except when my Linear Algebra professor went at 3x speed :sweat_smile:)

> I think so, Iirc postfix also work with `regTrig` (those two concepts are orthogonal: postfix just finds the word before the trigger, this happens entirely after the trigger is matched)

Makes sense. So would it be possible to make something with a trigger like `vr`, but then for words that have "vr" in them, they don't trigger?

> BTW, interesting, how so?

I guess I could do some trigger for all the math keywords, like `mm` or something, which appends the `\` needed to them. At least as a short term solution.

**Reply by L3MON4D3** [reply]
> Woops, school got the best of me this week, but at least taking notes was pretty easy with LuaSnips

That's what it's all for xD

> but then for words that have "vr" in them, they don't trigger?

That sounds like `wordTrig`: if it is set for some snippet, the trigger has to completely match the word before the cursor, and stuff like `abcdevr` wouldn't expand

> I guess I could do some trigger for all the math keywords, like mm or something, which appends the \ needed to them. At least as a short term solution.

Ahhh right, yeah that would work
Although a snippet sounds a bit overkill, maybe just put `\` behind some quick key-combo so it's easier to type?

**Comment by ejmastnak** [comment]
Late to the party here, but just finished a guide addressing exactly this UltiSnips-to-LuaSnip-in-a-LaTeX-context issue and thought I'd share it here: https://www.ejmastnak.com/tutorials/vim-latex/luasnip.html

(@evesdropper , I know you recently wrote [a guide of your own](https://github.com/evesdropper/dotfiles/tree/main/nvim/luasnip#readme), so I guess this is a bit redundant.)

@L3MON4D3, although the examples are LaTeX-flavored, do you think this might also be useful as a general reference for new users getting started with Lua snippets, LaTeX-oriented or otherwise? I did a similar sort of beginner's tutorial [over at VimTeX](https://github.com/lervag/vimtex#tutorial)---it seems like GIFs, highlighted code blocks and other such eye candy are well-received by new users even if all the info is already in the Wiki or docs haha

**Reply by evesdropper** [reply]
Looks cool :eyes:

Didn't think of making visual a separate function when I was working with it, will now implement in my snippets now. Will definitely save space and time!

Also, I believe instead of `autosnippet = true` you could also make a separate table. Probably down to personal preference I think.

Do tell how the GIFs are made? I've been dabbling with content creation for about 5 years and after everything I've learned I still have no idea how to create GIFs :laughing:

**Reply by L3MON4D3** [reply]
> Late to the party here, but just finished a guide addressing exactly this UltiSnips-to-LuaSnip-in-a-LaTeX-context issue and thought I'd share it here: https://www.ejmastnak.com/tutorials/vim-latex/luasnip.html

Oh niiice, that's very detailed :+1:

> @L3MON4D3, although the examples are LaTeX-flavored, do you think this might also be useful as a general reference for new users getting started with Lua snippets, LaTeX-oriented or otherwise? I did a similar sort of beginner's tutorial [over at VimTeX](https://github.com/lervag/vimtex#tutorial)---it seems like GIFs, highlighted code blocks and other such eye candy are well-received by new users even if all the info is already in the Wiki or docs haha

Definitely! The wiki is more of a reference, but not great to read front-to-back.
Feel free to add it to the README.
(Also +1 for gifs and eye candy :D)

**Reply by ejmastnak** [reply]
> Do tell how the GIFs are made? I've been dabbling with content creation for about 5 years and after everything I've learned I still have no idea how to create GIFs

The GIFs are made with [Menyoki](https://github.com/orhun/menyoki), if you're on an X11 Linux system I'd highly recommend it---it has great docs, does screenshot and screen record in a number of formats, and is scriptable/command-line friendly, if that's your kind of thing. So you might run something like

```bash
# Record root window, use the slop utility to select recording region with mouse, save in GIF format and encode with gifksi
menyoki record --root --size $(slop) gif --gifski save "${HOME}/Pictures/cap.gif"
```
The scriptability makes the process easy to automate, if you want to go down a rabbit hole you could check out the GIF-related source files at https://github.com/ejmastnak/ejmastnak.github.io/tree/main/tutorials/vim-latex/gifs/ultisnips; would be happy to share more if you're interested.

**Reply by ejmastnak** [reply]
> Definitely! The wiki is more of a reference, but not great to read front-to-back.
> Feel free to add it to the README.
> (Also +1 for gifs and eye candy :D)

Awesome, I'll do a PR later today or tomorrow!

**Reply by evesdropper** [reply]
> The GIFs are made with [Menyoki](https://github.com/orhun/menyoki), if you're on an X11 Linux system I'd highly recommend it---it has great docs, does screenshot and screen record in a number of formats, and is scriptable/command-line friendly, if that's your kind of thing. 

Honestly sounds super interesting - I'd definitely love to learn more. I've made a few scripts for custom screenshot behavior so hopefully some of that knowledge will carry over :sweat_smile: 

Oof, seems like no Wayland support though (rip me using Ubuntu 22.04 and Sway on Arch); however I am thinking of setting up an X11 based WM on Arch after exams. Looking forward to learning more!

**Comment by ejmastnak** [comment]
> Honestly sounds super interesting - I'd definitely love to learn more.

@evesdropper cool, I'll put together a write-up sometime and ping you when it's ready

**Comment by evesdropper** [comment]
A bit of a side tangent but would it also be a good idea to indicate in resources/readme or something that newer users can take advantage of this discussion for help on converting snippets just to centralize knowledge?

**Comment by hanschen** [comment]
A quick question: In Ultisnips I have snippets with repeated placeholders, for example:
```
snippet b
\begin{$1}
	$0
\end{$1}
endsnippet
```

I can use the same snippet (with some small modifications) with LuaSnip's snipmate loader, but when I fill in the first `$1`, the second `$1` in `\end{$1}` isn't updated until I hit `<Tab>` (to jump to the next placeholder). In Ultisnips, the second `$1` is updated concurrently as I fill in the first `$1` in insert mode.  

I prefer the Ultisnips behavior because I sometimes cancel the session (by leaving insert mode) after filling in the first `$1`. With LuaSnip I have to remember to hit `<Tab>` before I `<Esc>` to normal mode. (Also the Ultisnips behavior is visually nice, but that's just a minor point.)

**Reply by ejmastnak** [reply]
Does setting `update_events = 'TextChanged,TextChangedI'` solve this for you? For example:

```lua
ls.config.set_config({
  update_events = 'TextChanged,TextChangedI'
})
```

This should update all nodes and their dependencies on every keystroke in insert mode, i.e. concurrently as you fill in the first $1. Here are the docs if you're interested: https://github.com/L3MON4D3/LuaSnip/blob/master/DOC.md#config-reference

If this doesn't work, my next suggestion would be to write the snippet in pure Lua and using the `rep()` node.

**Reply by hanschen** [reply]
That does work, many thanks for the help. This makes LuaSnip much nicer to use for me.

I think it would be good to gather all such small tips in a central place (maybe under [Nice Configs](https://github.com/L3MON4D3/LuaSnip/wiki/Nice-Configs) in the wiki?) for users migrating from Ultisnips, as @evesdropper mentioned above. Here are some options I personally use:

1. Load textmate-style snippets with `require("luasnip.loaders.from_snipmate").lazy_load()` (https://github.com/L3MON4D3/LuaSnip#add-snippets)
2. Expand snippets in visual mode (replace `${VISUAL}` with `$TM_SELECTED_TEXT`) (https://github.com/L3MON4D3/LuaSnip/blob/master/DOC.md#snipmate)
3. Enable nested placeholders using `parser_nested_assembler` (https://github.com/L3MON4D3/LuaSnip/wiki/Nice-Configs#imitate-vscodes-behaviour-for-nested-placeholders)
4. Update (repeated) placeholders on every keystroke using `update_events` (https://github.com/L3MON4D3/LuaSnip/blob/master/DOC.md#config-reference)
5. Cancel snippet when leaving insert mode (I currently use this solution: https://github.com/L3MON4D3/LuaSnip/issues/656#issuecomment-1313310146)
6. Some people may want a "super-tab" functionality, which is mentioned in the README (link to https://github.com/hrsh7th/nvim-cmp/wiki/Example-mappings#luasnip)

**Reply by evesdropper** [reply]
:eyes: Yeah, a directory with a list of these small config setups seems like a great idea. If anything you could probably add them in the wiki yourself and maybe organize by the type of change you want in your config (e.g. setup with cmp, general snippet configs like adding snippets, more granular configs for specific type of snippets, visual/aesthetic/personal preference configs).

**Reply by L3MON4D3** [reply]
Yes! There's lots of very useful stuff scattered across `DOC.md` too (edit snippets by filetype, various nice additions to `snip_env`, `extend_decorator`). IMO there are enough to warrant a separate, opinionated repo which just provides a well-rounded luasnip-setup (either to just include in your config, or to steal ideas from :D), but an entry in the wiki sounds useful as well :+1:

**Reply by hanschen** [reply]
I've added a preliminary page here: https://github.com/L3MON4D3/LuaSnip/wiki/Migrating-from-UltiSnips

Feel free to add to it, improve the text, and/or move the information to a more suitable place.

**Comment by PorcelainMouse** [comment]
So, I'm not wrong in thinking that LuaSnip doesn't include any snips.  That was a unexpected.

Rather than trying to convert all UltiSnips, I found this repo that seems like  set of basic snips that you can add to LuaSnips: https://github.com/molleweide/LuaSnip-snippets.nvim
